`mousemove` is not fired on parent of `mouseover` targets that are removed at `mouseover`

lilyspiniolas opened the pull request:
#### 1651c7a400657b0b7d885fe0fb07d9bf869eafa0
<pre>
`mousemove` is not fired on parent of `mouseover` targets that are removed at `mouseover`
<a href="https://bugs.webkit.org/show_bug.cgi?id=299398">https://bugs.webkit.org/show_bug.cgi?id=299398</a>
<a href="https://rdar.apple.com/161203639">rdar://161203639</a>

Reviewed by Abrar Rahman Protyasha.

`EventHandler::dispatchMouseEvent` calls `updateMouseEventTargetNode` which fires
boundary events such as `mouseover` when necessary. If a page has event listeners which
delete the target node when `mouseover` fires for the target node, we will still attempt
to dispatch a `mousemove` event to the disconnected node.

To fix this, `EventHandler::nodeWillBeRemoved` now stores the ancestor of the node to be
removed if the node is the current element underneath the mouse. Then, in
`EventHandler::dispatchMouseEvent`, if the override is set and the event type is `mousemove`,
we dispatch the event to override element instead (the parent of the removed element).

Tested by imported/w3c/web-platform-tests/pointerevents/mousemove_after_mouseover_target_removed.html

* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_pointer_boundary_events_after_reappending_last_over_target-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointermove_after_pointerover_target_removed.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/mousemove_after_mouseover_target_removed-expected.txt:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::nodeWillBeRemoved):
(WebCore::EventHandler::updateMouseEventTargetAfterLayoutIfNeeded):
(WebCore::EventHandler::dispatchMouseEvent):
* Source/WebCore/page/EventHandler.h:

Canonical link: <a href="https://commits.webkit.org/300522@main">https://commits.webkit.org/300522@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9668a368195296010eabb3c6c29cff5f8da887d6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122792 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/42504 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33194 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129418 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/74896 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3b4e28b5-5e7a-419b-99c3-e35016b3d98f) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124668 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43223 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51097 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93321 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/61956 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e1eb6d1d-286f-4087-b2ac-132a5926f7fe) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125744 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34464 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109921 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73962 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/5ef358a4-4533-4403-a732-d43c8a502219) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33446 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28082 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72908 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104164 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28290 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132144 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49737 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37862 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101835 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50114 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106137 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101710 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25857 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47082 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25265 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/46507 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49594 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/55347 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49061 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/52529 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50744 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/055260e57c4cd33c2f57ddfabab00cdfc5b267b5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51226

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/526ebe1b217c98fc7a73cd144736d007fba85975. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51226

aprotyas commented on LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_pointer_boundary_events_after_reappending_last_over_target-expected.txt:
```diff
@@ -1,5 +1,5 @@
 
-FAIL After re-appending the last over element at pointerover, pointer boundary events should be fired on the original target again assert_equals: expected "pointerover@div#child, pointerenter@body, pointerenter@div#grandparent, pointerenter@div#parent, pointerenter@div#child, pointermove@div#child, pointerover@div#child, pointerenter@div#child, pointermove@div#child" but got "pointerover@div#chil ...
-FAIL After re-appending the last over element at pointerenter, pointer boundary events should be fired on the original target again assert_equals: expected "pointerover@div#child, pointerenter@body, pointerenter@div#grandparent, pointerenter@div#parent, pointerenter@div#child, pointermove@div#child, pointerover@div#child, pointerenter@div#child, pointermove@div#child" but got "pointerover@div#chi ...
+FAIL After re-appending the last over element at pointerover, pointer boundary events should be fired on the original target again assert_equals: expected "pointerover@div#child, pointerenter@body, pointerenter@div#grandparent, pointerenter@div#parent, pointerenter@div#child, pointermove@div#child, pointerover@div#child, pointerenter@div#child, pointermove@div#child" but got "pointerover@div#chil ...
+FAIL After re-appending the last over element at pointerenter, pointer boundary events should be fired on the original target again assert_equals: expected "pointerover@div#child, pointerenter@body, pointerenter@div#grandparent, pointerenter@div#parent, pointerenter@div#child, pointermove@div#child, pointerover@div#child, pointerenter@div#child, pointermove@div#child" but got "pointerover@div#chi ...
```
These may be easy candidates for a coming progression? We fire an unnecessary `pointermove@div#parent`

aprotyas reviewed (approved):
Small typo in the commit message:

> dispatch the _event_ to override element

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9668a368195296010eabb3c6c29cff5f8da887d6. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51226

lilyspiniolas committed:
`mousemove` is not fired on parent of `mouseover` targets that are removed at `mouseover`
https://bugs.webkit.org/show_bug.cgi?id=299398
rdar://161203639

Reviewed by Abrar Rahman Protyasha.

`EventHandler::dispatchMouseEvent` calls `updateMouseEventTargetNode` which fires
boundary events such as `mouseover` when necessary. If a page has event listeners which
delete the target node when `mouseover` fires for the target node, we will still attempt
to dispatch a `mousemove` event to the disconnected node.

To fix this, `EventHandler::nodeWillBeRemoved` now stores the ancestor of the node to be
removed if the node is the current element underneath the mouse. Then, in
`EventHandler::dispatchMouseEvent`, if the override is set and the event type is `mousemove`,
we dispatch the event to override element instead (the parent of the removed element).

Tested by imported/w3c/web-platform-tests/pointerevents/mousemove_after_mouseover_target_removed.html

* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointerevent_pointer_boundary_events_after_reappending_last_over_target-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/pointerevents/pointermove_after_pointerover_target_removed.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/uievents/mouse/mousemove_after_mouseover_target_removed-expected.txt:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::nodeWillBeRemoved):
(WebCore::EventHandler::updateMouseEventTargetAfterLayoutIfNeeded):
(WebCore::EventHandler::dispatchMouseEvent):
* Source/WebCore/page/EventHandler.h:

Canonical link: https://commits.webkit.org/300522@main
